core/src/main/java/io/questdb/cutlass/http/processors/LineHttpProcessorState.java

```diff
@@ -0,0 +1,446 @@
+    private long errorId;
+    private int fd = -1;
+    private int line = 0;
+    private long recvBufEnd;
+    private long recvBufPos;
+    private long recvBufStartOfMeasurement;
+    private SecurityContext securityContext;
+    private SendStatus sendStatus = SendStatus.NONE;
+
+    public LineHttpProcessorState(int recvBufSize, int maxResponseContentLength, CairoEngine engine, LineHttpProcessorConfiguration configuration) {
+        assert recvBufSize > 0;
+        this.recvBufSize = recvBufSize;
+
+        assert HttpServerConfiguration.MIN_SEND_BUFFER_SIZE > 100;
+        // Response is measured in bytes some error messages can have non-ascii characters
+        // approximate 1.5 bytes per character
+        this.maxResponseErrorMessageLength = (int) ((maxResponseContentLength - 100) / 1.5);
+        this.recvBufPos = this.buffer = Unsafe.malloc(recvBufSize, MemoryTag.NATIVE_HTTP_CONN);
+        this.recvBufEnd = this.recvBufPos + recvBufSize;
+        this.parser = new LineTcpParser(configuration.isStringAsTagSupported(), configuration.isSymbolAsFieldSupported());
+        this.parser.of(buffer);
+        this.appender = new IlpWalAppender(configuration.autoCreateNewColumns(),
+                configuration.isStringToCharCastAllowed(),
+                configuration.getTimestampAdapter(),
+                engine.getConfiguration().getMaxFileNameLength(),
+                configuration.getMicrosecondClock()
+        );
+        DefaultColumnTypes defaultColumnTypes = new DefaultColumnTypes(configuration.getDefaultColumnTypeForFloat(), configuration.getDefaultColumnTypeForInteger());
+        this.ilpTudCache = new IlpTudCache(
+                engine,
+                configuration.autoCreateNewColumns(),
+                configuration.autoCreateNewTables(),
+                defaultColumnTypes,
+                configuration.getDefaultPartitionBy()
+        );
+        symbolCachePool = new WeakClosableObjectPool<>(
+                () -> new SymbolCache(configuration.getMicrosecondClock(), configuration.getSymbolCacheWaitUsBeforeReload()), 5);
+    }
+
+    public void clear() {
+        ilpTudCache.clear();
+        Vect.memset(buffer, recvBufSize, 0);
+        parser.of(buffer);
+        recvBufPos = buffer;
+        error.clear();
+        currentStatus = Status.OK;
+        errorLine = 0;
+        line = 0;
+        recvBufStartOfMeasurement = 0;
+        sendStatus = SendStatus.NONE;
+    }
+
+    @Override
+    public void close() {
+        Unsafe.free(buffer, recvBufSize, MemoryTag.NATIVE_HTTP_CONN);
+        recvBufStartOfMeasurement = recvBufEnd = recvBufPos = buffer = 0;
+        Misc.free(ilpTudCache);
+        Misc.free(symbolCachePool);
+    }
+
+    public void commit() {
+        try {
+            ilpTudCache.commitAll();
+        } catch (Throwable th) {
+            ilpTudCache.setDistressed();
+            currentStatus = handleCommitError(th);
+        }
+    }
+
+    public void formatError(Utf8Sink sink) {
```

Comment:
`LineHttpProcessor` sets content type header for errors as `text/plain`, but here we are constructing a JSON-like message. Is that intentional? (It may result in an invalid JSON, depending on error/errorLine content, but that's probably unrelated?)

Reply:
JSON should be valid, `sink.put()` encodes JSON.
I will change the response encoding in the header.

Reply:
done in [1380437](https://github.com/questdb/questdb/pull/4034/commits/13804375a595e586fde3ef559ac83e92e9da2e1b)